pybamm/spatial_methods/finite_volume.py

```diff
@@ -234,21 +271,17 @@ def divergence_matrix(self, domain, bc_type="dirichlet"):
 
         # Create matrix using submesh
         n = submesh.npts + 1
-        if bc_type == "dirichlet":
-            data = np.vstack(
-                [
-                    np.concatenate([-e, np.array([0])]),
-                    np.concatenate([np.array([0]), e]),
-                ]
-            )
-            diags = np.array([0, 1])
-            sub_matrix = spdiags(data, diags, n - 1, n)
-        elif bc_type == "neumann":
-            # we don't have to act on bc fluxes which are now in
-            # the bc vector
-            data = np.vstack([-e[1:], e[:-1]])
-            diags = np.array([-1, 0])
-            sub_matrix = spdiags(data, diags, n - 1, n - 2)
+        if bc_type == "dirichlet_dirichlet":
+            sub_matrix = diags([-e, e], [0, 1], shape=(n - 1, n))
+        elif bc_type == "dirichlet_neumann":
+            # we don't have to act on right bc flux which is now in the bc vector
+            sub_matrix = diags([-e, e], [0, 1], shape=(n - 1, n - 1))
```

Comment:
You probably have already checked this but just want to say that we should be really careful to test each of these submatrix forms as results really depend on having the correct vectors e[1:] or e. If we don't the 1/dx terms are associated with the wrong variables. At the moment this will just appear as errors at the neg_sep and sep_pos interfaces.